app/services/user_stat_count_updater.rb

```diff
@@ -3,16 +3,27 @@
 class UserStatCountUpdater
   class << self
     def increment!(post, user_stat: nil)
-      update!(post, user_stat: user_stat)
+      update_using_operator!(post, user_stat: user_stat, action: :increment!)
     end
 
     def decrement!(post, user_stat: nil)
-      update!(post, user_stat: user_stat, action: :decrement!)
+      update_using_operator!(post, user_stat: user_stat, action: :decrement!)
+    end
+
+    def set!(user_stat:, count:, count_column:)
+      return if user_stat.blank?
+      return if ![:post_count, :topic_count].include?(count_column)
+
+      if SiteSetting.verbose_user_stat_count_logging && count < 0
+        Rails.logger.warn("Attempted to insert negative count into UserStat##{count_column} for user #{user_stat.user_id}, using 0 instead.")
```

Comment:
I wonder if we should improve our logging and add the name of the method that called `set!`?